Add setting UI components to the verify screen

Naxesss opened the pull request:
Split off from https://github.com/ppy/osu/pull/12751, related to https://github.com/ppy/osu/issues/12091.

Reorganizes a few UI components and adds new ones:
- Added sections to `IssueSettings`, all inheriting a new `Section` class:
    - `VisibilitySection` for filtering issue types
        - Warnings and errors can be toggled
        - Negligible issues are hidden by default
    - `InterpretationSection` for adjusting interpreted difficulty level (does nothing yet)
- Factored out `Settings` from `ControlPointSettings` for use in `IssueSettings`
- Separated `IssueList` from `IssueTable`, as it's now also used for setting bindables

https://user-images.githubusercontent.com/30292137/117898397-2d1b5700-b2c5-11eb-99f2-f81885d6f5a4.mp4

---

Did what @PercyDan54 suggested in the other PR, and am no longer refactoring `ControlPointSection` because of that.

Naxesss committed:
Factor out `load` from settings into new `Settings` class

Naxesss committed:
Take `IssueList` in `IssueSettings` constructor

We'll be using this for bindables later.

Naxesss committed:
Separate `IssueList` into own class

Naxesss committed:
Add abstract `Section` class

Similar to `Section` in the timing screen, but does not make use of checkboxes, nor specific to control points.

So there's a lot of things that differ, hence new class instead of factoring that out.

Naxesss committed:
Add `VisibilitySection` and its bindables in `IssueList`

Naxesss committed:
Add `IssueList` filtering based on those bindables

Naxesss committed:
Refresh `IssueList` on changes in `VisibilitySection`

Naxesss committed:
Remove negligible default hidden TODO

Naxesss committed:
Add `InterpretationSection` and its bindable in `IssueList`

We'll eventually connect that bindable so that checks can access it.

peppy committed:
Rename `Settings` to have a more localised name

peppy committed:
Reverse direction of binding to allow for better abstract class definitions

peppy commented:
I've done an initial clean-up pass on this which renames classes to avoid over-generic naming like `Settings`. I've also changed some of the `ctor` passing of data down to instead use upwards binding to the `VerifyScreen` itself (this allows for the sections to be reused without worrying about `ctor` requirements, and should lead to better testability too).

One remaining case is `VisibilitySection.ShowType`, which should probably also be moved, but I'd want to see it hopefully implemented as a `BindableList<IssueType>` rather than the dictionary you are currently using. I'll take a look at this in another pass unless you beat me to it.

Naxesss committed:
Move `ShowIssueTypes` to `VerifyScreen`

Naxesss committed:
Remove unnecessary local variable

This now exists in `VerifyScreen`, which we can access from here.

Naxesss committed:
Set interpreted difficulty to correct default

Naxesss committed:
Split `ShowIssueTypes` dict into hidden and configurable lists

This way `VerifyScreen` is decoupled from which options `VisibilitySection` provides.

Bindings are a bit less neat, though.

peppy commented on osu.Game/Screens/Edit/Verify/VisibilitySection.cs:
```diff
@@ -28,8 +35,18 @@ private void load(OverlayColourProvider colours)
+                    else
+                        verify.HiddenIssueTypes.Remove(issueType);
+
+                    verify.IssueList.Refresh();
```
This should be moved to inside `IssueList.cs` itself using something like

```csharp
HiddenIssueTypes.BindCollectionChanged(_ => Refresh(), true);
```

peppy commented on osu.Game/Screens/Edit/Verify/VisibilitySection.cs:
```diff
@@ -28,8 +35,18 @@ private void load(OverlayColourProvider colours)
 
-                checkbox.Current.BindTo(verify.ShowIssueType[issueType]);
-                checkbox.Current.BindValueChanged(_ => verify.IssueList.Refresh());
+                checkbox.Current.Default = !verify.HiddenIssueTypes.Contains(issueType);
```
You'll also likely want to create a local bound copy to avoid cross-component bindings:

```csharp
hiddenIssueTypes = verify.HiddenIssueTypes.GetBoundCopy();
```

and use that field in this class rather than direct references to `verify.`.

Naxesss replied:
Out of curiousness, are cross-component bindings are mainly an issue due to the bindings becoming unorganized and more difficult to debug/trace, or am I missing the point?

peppy replied:
To give a very brief history lesson, `Bindable`s are our solution to event/memory leaks caused by using the classic `event` binding pattern (ie. `ThingChanged += ourThing` and forgetting to remove the binding). To make this work properly, each class needs to maintain their own instance.

If this rule is broken, as an example, your `VisibilitySection` is holding a reference to its parent (because you are binding a local method with implicit references to fields of the instance). While generally this isn't an issue because you presume the parent is cleaned up along with its children, it is a larger issue in the opposite direction. Imagine doing something like `game.CurrentScreen.BindValueChanged(screen => { this.currentScreenText.Text = $"{screen}"; })`. Your local instance would now be held for the full duration of game execution.

For the sake of good practices we tend to avoid this in both directions.

There's one more reason: `Drawable` has code to automatically unbind all bindable fields/properties when an instance goes out of context. This stops events from firing at all, where firing them could lead to unexpected behaviour, performance implications, or cross-thread exceptions.

Naxesss replied:
That makes sense. I've applied the same change to `InterpretationSection`, as I figure the same thing applies there (and in doing so I also realized I forgot to bind it to refresh on update, so did that in `IssueList` like with `HiddenIssueTypes`).

Naxesss committed:
Isolate refreshing to `IssueList`

Naxesss committed:
Use local bound copy for `HiddenIssueTypes`

peppy commented on osu.Game/Screens/Edit/Verify/VisibilitySection.cs:
```diff
@@ -26,6 +26,8 @@ internal class VisibilitySection : EditorRoundedScreenSettingsSection
         [BackgroundDependencyLoader]
         private void load(OverlayColourProvider colours)
         {
+            var hiddenIssueTypes = verify.HiddenIssueTypes.GetBoundCopy();
```
you'll have to store this to a field. bindables which don't have a reference are automatically cleaned up (all bindings are tracked using weak references internally).

peppy commented on osu.Game/Screens/Edit/Verify/IssueList.cs:
```diff
@@ -83,10 +83,10 @@ protected override void LoadComplete()
             base.LoadComplete();
 
-            Refresh();
+            verify.HiddenIssueTypes.BindCollectionChanged((o, s) => refresh(), runOnceImmediately: true);
```
```suggestion
            verify.HiddenIssueTypes.BindCollectionChanged((o, s) => refresh(), true);
```

just to match how we do it everywhere else.

Naxesss committed:
Use more consistent lambda discards

Naxesss committed:
Keep track of local bound copy

Naxesss committed:
Use local bound copy in `InterpretationSection` as well

Else we're relying on the `VerifyScreen`'s bindable instance, and by extension the `VerifyScreen` instance itself.

Naxesss committed:
Also refresh when interpreted difficulty changes

Naxesss committed:
Set default for bindable in object initializer

Fixes the CI failure.

peppy committed:
Tidy up specification of `SettingsSection`

peppy committed:
Simplify binding flow in `InterpretationSection`

peppy commented on osu.Game/Screens/Edit/Verify/VisibilitySection.cs:
```diff
@@ -0,0 +1,57 @@
+                    Anchor = Anchor.CentreLeft,
+                    Origin = Anchor.CentreLeft,
+                    LabelText = issueType.ToString(),
+                    Current = { Default = !hiddenIssueTypes.Contains(issueType) }
```
Worth noting that if we ever add saving support for the user settings, this will break. But seems fine for now.

peppy committed:
Move DI resolution to inside BDL parameters
